tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs

```diff
@@ -4,22 +4,25 @@
 // </copyright>
 
 using System;
-using Datadog.Trace.Configuration;
-using Datadog.Trace.Util.Http.QueryStringObfuscation;
+using Datadog.Trace.Tagging;
 
 namespace Datadog.Trace.Util.Http
 {
-    internal class HttpRequestUtils
+    internal static class HttpRequestUtils
     {
         private const string NoHostSpecified = "UNKNOWN_HOST";
 
-        internal static string GetUrl(string scheme, string host, string pathBase, string path, string queryString, QueryStringManager queryStringManager = null)
+        internal static LazyOrString GetUrl(string scheme, string host, string pathBase, string path, string queryString, QueryStringManager queryStringManager = null)
         {
-            if (queryStringManager != null)
+            if (queryStringManager?.ShouldObfuscate(queryString) ?? false)
             {
-                queryString = queryString.Substring(0, Math.Min(queryString.Length, 200));
-                queryString = queryStringManager.Obfuscate(queryString);
-                return $"{scheme}://{(string.IsNullOrEmpty(host) ? NoHostSpecified : host)}{pathBase}{path}{queryString}";
+                return new(
+                    () =>
+                    {
+                        var queryStringLocal = queryString.Substring(0, Math.Min(queryString.Length, 200));
```

Comment:
We should probably check if `queryString.Length < 200` before calling `queryString.Substring()`. Even though there will be no heap allocations if the string is shorter than 200 (`Substring()` will return the same string reference), there's some bounds checking inside `Substring()` than we can skip, not to mention the method call itself.

```suggestion
                        var queryStringLocal = queryString.Length < 200 ? queryString : queryString.Substring(0, 200);
```